### ISSUE TITLE
System/out should output on nREPL client

### ISSUE THREAD

**Issue by ivarref** [nrepl/nrepl#119]
## Expected behavior
```clojure
(.println System/out "Hello")
```

Should print on nREPL client.

## Actual behavior
Prints on nREPL server.

## Steps to reproduce the problem
```bash
# In terminal one:
$ clj -Sdeps '{:deps {nrepl {:mvn/version "0.5.3"}}}' -m nrepl.cmdline --port 5600 --headless
nREPL server started on port 5600 on host localhost - nrepl://localhost:5600

# In terminal two:
$ clj -Sdeps '{:deps {nrepl {:mvn/version "0.5.3"}}}' -m nrepl.cmdline --connect --host localhost --port 5600
user=> (.println System/out "Hello") ; Hello is written in server (terminal one).
nil
user=> 
```

Seems issue #29 notes a similar problem.

This broke Java logging for my project, i.e. the logs were written on the server, and not the client.

[This gist solved the problem](https://gist.github.com/daviesian/4517859) for me, but I think this fix (or a similar method) should be a part of nREPL's default behaviour.

Thanks and kind regards.

### Clojure version
1.9.0

### Java version
OpenJDK 64-Bit Server VM 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12

### Operating system
Ubuntu 18.04.1 LTS

**Comment by bbatsov** [comment]
Yeah, I completely agree that we should try harder to gather all output.

**Comment by cichli** [comment]
One question is at what time we should capture the values of `System/out` and `System/err`. Load-time? When the server starts? If other tools (e.g. log4j) also call `setOut` and `setErr` this might be important! Likewise if we `alter-var-root` either `*out*` or `*err*`. The CIDER implementation of this mutates all four every time a new client subscribes (see [here](https://github.com/clojure-emacs/cider-nrepl/blob/master/src/cider/nrepl/middleware/out.clj#L104-L110)) – I think we should aim to do this once and only once.

Load-time might be the best option since we do have [config](https://github.com/nrepl/nrepl/blob/master/src/clojure/nrepl/config.clj#L48-L56) available then. I think if we're going to hijack globals like `System/out` and `*out*` then there should be a way to completely turn the feature off in config.

**Comment by PEZ** [comment]
If this is fixed some day, I want to request that these messages are made special in some way. That way the nREPL client can choose to relay them in some special way. Some servers are quite chatty and the user might not always want this output appear the same way as stdout from immediately evaluated expressions do.

**Comment by bbatsov** [comment]
It's hard for the server to tell apart the different output messages. I was thinking back in the day that maybe they can be processed differently by clients based on regular expressions, or something along those lines, but I recall I hit a wall when working in that direction in CIDER.

**Comment by ivarref** [comment]
I've done some work towards this at: https://github.com/ivarref/nrepl

Currently missing configuration options, linting, tests, etc.. But seems to work OK IMO.

> If this is fixed some day, I want to request that these messages are made special in some way. That way the nREPL client can choose to relay them in some special way. Some servers are quite chatty and the user might not always want this output appear the same way as stdout from immediately evaluated expressions do.

In my fork lines are currently sent like this:
```
{:session sess-id
 :out (str line "\n")
 :source :redirect}
```

I think it makes sense to make this feature opt-in, i.e. it's disabled by default?

**Comment by bbatsov** [comment]
@ivarref I agree. Let's make this opt-in.
